src/runtime/security/sasl_utils.cpp

```diff
@@ -0,0 +1,157 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "sasl_utils.h"
+#include "kinit_context.h"
+
+#include <sasl/sasl.h>
+#include <sasl/saslplug.h>
+#include <functional>
+
+#include <dsn/c/api_utilities.h>
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/utility/flags.h>
+#include <dsn/utility/synchronize.h>
+
+namespace dsn {
+namespace security {
+DSN_DEFINE_string("security", sasl_plugin_path, "/usr/lib/sasl2", "path to search sasl plugins");
+
+const char *logger_level_to_string(int level)
+{
+    switch (level) {
+    case SASL_LOG_NONE:
+        return "SASL_LOG_NONE";
+    case SASL_LOG_ERR:
+        return "SASL_LOG_ERR";
+    case SASL_LOG_FAIL:
+        return "SASL_LOG_FAIL";
+    case SASL_LOG_WARN:
+        return "SASL_LOG_WARN";
+    case SASL_LOG_NOTE:
+        return "SASL_LOG_NOTE";
+    case SASL_LOG_DEBUG:
+        return "SASL_LOG_DEBUG";
+    case SASL_LOG_TRACE:
+        return "SASL_LOG_TRACE";
+    case SASL_LOG_PASS:
+        return "SASL_LOG_PASS";
+    default:
+        return "Unkown SASL log level";
+    }
+}
+
+int sasl_simple_logger(void *context, int level, const char *msg)
+{
+    if (SASL_LOG_NONE == level || nullptr == msg) {
+        return SASL_OK;
+    }
+
+    ddebug_f("sasl log info: log level = {}, message = {}", logger_level_to_string(level), msg);
```

Comment:
All logs are in DEBUG level? I think it;s needed to convert `level` to related rdsn's log level.

Reply:
I have thought about this. But the log level in sasl is not a one-to-one correspondence log level in rdsn.
log level in rdsn:
```
typedef enum dsn_log_level_t {
    LOG_LEVEL_INFORMATION,
    LOG_LEVEL_DEBUG,
    LOG_LEVEL_WARNING,
    LOG_LEVEL_ERROR,
    LOG_LEVEL_FATAL,
    LOG_LEVEL_COUNT,
    LOG_LEVEL_INVALID
} dsn_log_level_t;
```
log level in sasl:
```
define SASL_LOG_NONE  0	/* don't log anything */
#define SASL_LOG_ERR   1	/* log unusual errors (default) */
#define SASL_LOG_FAIL  2	/* log all authentication failures */
#define SASL_LOG_WARN  3	/* log non-fatal warnings */
#define SASL_LOG_NOTE  4	/* more verbose than LOG_WARN */
#define SASL_LOG_DEBUG 5	/* more verbose than LOG_NOTE */
#define SASL_LOG_TRACE 6	/* traces of internal protocols */
#define SASL_LOG_PASS  7	/* traces of internal protocols, including
```

Reply:
How about?
```
SASL_LOG_ERR -> LOG_LEVEL_ERROR
SASL_LOG_FAIL/SASL_LOG_WARN -> LOG_LEVEL_WARNING
SASL_LOG_NOTE -> LOG_LEVEL_INFORMATION
othres -> LOG_LEVEL_DEBUG
```

Reply:
```
SASL_LOG_ERR -> LOG_LEVEL_ERROR
SASL_LOG_FAIL/SASL_LOG_WARN -> LOG_LEVEL_WARNING
SASL_LOG_NOTE -> LOG_LEVEL_DEBUG
othres -> LOG_LEVEL_INFORMATION

```